lib/screens/home/view_more/hidden_proposal_creation/proposal_creation_bloc.dart

```diff
@@ -0,0 +1,151 @@
+          isLoading: true,
+        ),
+      );
+    }
+
+    try {
+      final asset =
+          (await get<AssetService>().getAssets(account.coin, account.address))
+              .firstWhere((element) => element.denom == nHashDenom);
+
+      _creationDetails.tryAdd(
+        ProposalCreationDetails.fromExisting(
+          oldDetails,
+          hashAmount: nHashToHash(BigInt.parse(asset.amount)).toDouble(),
+        ),
+      );
+    } finally {
+      if (showLoading) {
+        final newerDetatils = _creationDetails.value;
+        _creationDetails.tryAdd(
+          ProposalCreationDetails.fromExisting(
+            newerDetatils,
+            isLoading: false,
+          ),
+        );
+      }
+    }
+  }
+
+  @override
+  gov.MsgSubmitProposal getMessage() {
+    final details = _creationDetails.value;
+    final scaledAmount =
+        hashToNHash(Decimal.parse(details.initialDeposit.toInt().toString()))
+            .toString();
+
+    return gov.MsgSubmitProposal(
+      content: gov.TextProposal(
+        title: details.title,
+        description: details.description,
+      ).toAny(),
+      initialDeposit: details.initialDeposit.toInt() > 0
+          ? [
+              Coin(
+                denom: nHashDenom,
+                amount: scaledAmount,
+              )
+            ]
+          : null,
+      proposer: account.address,
+    );
+  }
+
+  void updateTitle(String text) {
+    final oldDetatils = _creationDetails.value;
+    _creationDetails.tryAdd(
+      ProposalCreationDetails.fromExisting(
+        oldDetatils,
+        title: text,
+      ),
+    );
+  }
+
+  void updateDescription(String text) {
+    final oldDetatils = _creationDetails.value;
+    _creationDetails.tryAdd(
+      ProposalCreationDetails.fromExisting(
+        oldDetatils,
+        description: text,
+      ),
+    );
+  }
+
+  void updateInitialDeposit(double deposit) {
+    final oldDetatils = _creationDetails.value;
+    _creationDetails.tryAdd(
+      ProposalCreationDetails.fromExisting(
+        oldDetatils,
+        initialDeposit: deposit,
+      ),
+    );
+  }
+
+  @override
+  FutureOr onDispose() {
+    _creationDetails.close();
+  }
+}
+
+class ProposalCreationDetails {
+  ProposalCreationDetails({
+    this.isLoading = true,
+    this.initialDeposit = 0,
+    this.hashAmount = 0,
+    this.title = "",
+    this.description = "",
+  });
+
+  static ProposalCreationDetails fromExisting(
+    ProposalCreationDetails existing, {
+    bool? isLoading,
+    double? initialDeposit,
+    double? hashAmount,
+    String? title,
+    String? description,
+  }) {
+    return ProposalCreationDetails(
+      initialDeposit: initialDeposit ?? existing.initialDeposit,
+      description: description ?? existing.description,
+      hashAmount: hashAmount ?? existing.hashAmount,
+      isLoading: isLoading ?? existing.isLoading,
+      title: title ?? existing.title,
+    );
+  }
+
+  final bool isLoading;
+  final double initialDeposit;
```

Comment:
Perhaps good to be in the habit of using `Decimal` instead of `double` as it doesn't lose precision.
I think nhash should be in `BigInt`, and usd/hash others in `Decimal`.
When displayed to the user, I think we would then always truncate to a specified precision.

Reply:
Good point.